### PR TITLE
AirfRANS: 4L/256d + T_max=10 extended — prevent ep208 divergence

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current best (PR #2755, 0.003904 at ep201) used T_max=5 and diverged catastrophically at epoch 208 — gradient norms exploded to infinity and surface_mse jumped from 0.003904 to 0.628. Without this divergence, the model could have continued descending well below 0.003.

**Hypothesis:** T_max=5 creates aggressive cosine LR cycling (LR oscillates widely, peaking every 5 epochs) that eventually triggers irreversible gradient explosions in extended training. T_max=10 (slower cycling) should reduce the amplitude of LR peaks, preventing the catastrophic divergence while still enabling deep basin exploration.

Supporting evidence: The pressure-weight breakthrough (nami, 0.00435) used T_max=10 and achieved its best at ep117 with no divergence reported. T_max=10 may be more stable for extended runs.

If T_max=10 survives past ep208 and continues descending, the model should push below 0.003.

## Instructions

No code changes needed. Same golden config but with `--cosine-t-max 10`:

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 7e-4 \
  --cosine-t-max 10 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --agent shoya \
  --wandb-name "shoya/airfrans-4L256d-tmax10-extended" \
  --wandb-group "airfrans-tmax-stability"
```

**Key things to monitor:**
1. Does the model survive past epoch 208 without gradient explosion?
2. Best val_primary/surface_mse and at which epoch
3. How does convergence speed compare to T_max=5? (At ep56, T_max=5 was ~0.007; is T_max=10 faster or slower?)
4. Grad norm behavior around ep200-220

## Baseline

- **Current best:** val_primary/surface_mse = **0.003904** (PR #2755, shoya — your own previous run!)
- **Architecture:** 4L/256d + golden config (gc=1.0, WD=1e-2, T_max=5, AdamW lr=7e-4, 223 epochs)
- **W&B run:** stxm16tv
- **Diverged at:** ep208 (grad norm → infinity; T_max=5 aggressive cycling caused irreversible instability)
- **External target:** 0.0043 (beaten — now pushing below 0.003)
- **Reproduce:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shoya:

## Results

### Best Result
- **val_primary/surface_mse: 0.002780** at epoch 147 (**-28.8% vs baseline 0.003904**)
- Surface MSE breakdown (ep147): Ux=7.66e-05, Uy=3.69e-05, nut=8.34e-06, p=0.01100
- full_val/volume_mse: 0.02208
- Grad norm at best: 5.39

### Test Metrics (UNRELIABLE — evaluated on final diverged model, not best checkpoint)
- full_test/surface_mse: 0.15238 (final ep222, post-divergence)

### Divergence Analysis
**T_max=10 did NOT prevent divergence.** The model diverged at ep155-158, **52 epochs EARLIER** than T_max=5's ep208:

| Epoch | surface_mse | grad_norm | Status |
|-------|-------------|-----------|--------|
| 147 | 0.002780 | 5.39 | **BEST** |
| 154 | 0.005334 | 10.74 | Normal |
| 155 | 0.013447 | **43.38** | ⚠️ Spike |
| 156 | 0.098965 | **101.79** | 🔴 Exploding |
| 157 | 0.326459 | **1974.84** | 🔴 Catastrophic |
| 158+ | 0.240→0.181 | **inf** | Diverged plateau |

Post-divergence, the model plateaued at ~0.18-0.24 (vs T_max=5 going to complete infinity). Very slow descent from plateau but never recovered to useful levels.

### Convergence Speed Comparison (T_max=10 vs T_max=5)

| Milestone | T_max=10 (this run) | T_max=5 (PR #2755) |
|-----------|-------------------|-------------------|
| First < 0.01 | ep60 | ~ep43 |
| First < 0.005 | ep109 | ~ep158 |
| First < 0.004 | ep137 | ep201 |
| First < 0.003 | **ep147** | Never |
| Best achieved | **0.00278 (ep147)** | 0.003904 (ep201) |
| Diverged at | **ep155** | ep208 |

T_max=10 was slower in early epochs (ep56: 0.016 vs ~0.007 for T_max=5) but found deeper basins in the ep90-147 range. Trough envelope progression: 0.124 → 0.058 → 0.032 → 0.014 → 0.0087 → 0.0057 → 0.0038 → 0.00278.

### Key Monitoring Questions (from PR)

1. **Does the model survive past epoch 208?** — No. Diverged at ep155-158, 52 epochs earlier. However, the divergence character is different: T_max=10 plateaued at ~0.18 instead of going to complete infinity like T_max=5.

2. **Best val_primary/surface_mse and epoch?** — **0.002780 at ep147** (-28.8% vs baseline 0.003904, matches gc=0.5 result of 0.00277)

3. **Convergence speed vs T_max=5?** — Slower in early epochs (ep56: 0.016 vs ~0.007), but much deeper in later epochs. T_max=10 found sub-0.003 basins that T_max=5 never reached.

4. **Grad norm behavior ep200-220?** — Grad norms were already at infinity by ep158. At ep200-220 they remained at infinity but the model slowly descended from the divergence plateau (0.195→0.181), not recovering.

### Training Details
- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999 --agent shoya --wandb-name "shoya/airfrans-4L256d-tmax10-extended" --wandb-group "airfrans-tmax-stability"`
- **Epochs:** 222 (180-min timeout)
- **Peak memory:** ~54 GB VRAM
- **W&B run:** r303dhhy

### What happened
The hypothesis was partially confirmed and partially falsified:

**Confirmed:** T_max=10 produces deeper basins than T_max=5. The best result (0.00278 at ep147) is 28.8% better than T_max=5's best (0.003904 at ep201). T_max=10's slower cycling allows more thorough basin exploration during the descent phase.

**Falsified:** T_max=10 does NOT prevent divergence. In fact, it diverged 52 epochs earlier (ep155 vs ep208). The gradient explosion pattern is identical: grad norms spike from ~5 → 43 → 102 → 1975 → infinity in just 4 epochs. The root cause of divergence is not the cycling speed (T_max) but likely something structural — possibly the interaction between cosine LR restarts and the accumulated gradient landscape. The grad-clip=1.0 is insufficient to prevent the explosion once it starts.

### Suggested follow-ups
1. **Gradient clip 0.5 + T_max=10:** The gc=0.5 experiment (PR #2774) achieved 0.00277 without diverging. Combining gc=0.5 (proven stability) with T_max=10 (deeper basins) could push below 0.002.
2. **Warmup restarts:** Instead of hard cosine restarts, use a warmup phase after each restart to prevent the LR jump from triggering gradient explosions.
3. **Adaptive gradient clipping:** Scale clip threshold based on running average of grad norms — tighten clipping when norms start trending upward.
4. **Checkpoint-aware early stopping:** Since both T_max values eventually diverge, implement automatic early stopping when grad norms exceed a threshold (e.g., 20) for consecutive epochs.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2817 because ADVISOR: Closing — your 4L/256d T_max=10 result of 0.002780 does not beat the 0.001479 baseline. The 3L/256d architecture has proven decisively superior. Thank you for the thorough run, shoya..
